text/0148-logs-for-crashes.md

```diff
@@ -0,0 +1,222 @@
+3. Remove the log from the FIFO queue.
+4. If the queue isn’t empty, go to step 2.
+
+The FIFO queue has a `max-logs-count` of 64 logs. When the FIFO queue exceeds `max-logs-count` log items, the BatchProcessor MUST drop logs and record client reports with the category `queue_overflow` for every dropped log. SDKs MAY choose a different `max-logs-count` value, if needed.
+
+When a crash occurs, the SDKs write the logs in the FIFO queue to the `log-crash-recover-file`  and send these logs on the next SDK launch. To avoid sending duplicated logs, the SDKs MUST deduplicate the logs from the `log-crash-recover-file` and `batch-processor-cache-file` based on the logID after a crash before sending them. It's worth noting that logs don't have an ID yet, and we MUST extend the protocol for this.
+
+For the Cocoa SDK, the crash-safe memory space structure MUST be C memory, because Swift and Objective-C aren’t async-safe. For Java, it’s Java memory, because the JVM allows you to store logs in memory or on disk when a crash occurs. This solution also works for watchdog terminations, as the BatchProcessor MUST check on SDK launch if there are logs in the `batch-processor-cache-file` and send these.
+
+The BatchProcessor MUST keep two `batch-processor-cache-files`. When it sends the logs from `batch-processor-cache-file`, it renames it to `batch-processor-cache-file-to-flush` and creates a new `batch-processor-cache-file` to avoid losing logs if a crash occurs when flushing the logs. To avoid sending duplicate logs if the app crashes in between storing the envelope and deleting the cache file, the BatchProcessor MUST first store the envelope to the same folder as the BatchProcessor files. After deleting the `batch-processor-cache-file-to-flush`, SDKs MUST move the envelope to the envelope cache folder. As moving files can be done atomic, SDKs avoid sending duplicated logs in the described scenario. We're going to add a more detailed explanation once we add this concept to the develop docs. These are the flushing steps:
+
+1. Rename `batch-processor-cache-file` to `batch-processor-cache-file-to-flush`.
+2. Create a new `batch-processor-cache-file`
+3. Store new logs to `batch-processor-cache-file`
+4. Load logs into memory from the `batch-processor-cache-file-to-flush`
+5. Store the logs in memory into an envelope into the batch processor directory, which MUST be the same folder as the other BatchProcessor files, and name it `batch-processor-cache-envelope-to-flush`.
+6. Delete all logs from `batch-processor-cache-file-to-flush`
+7. Move the `batch-processor-cache-envelope-to-flush` to the envelopes cache folder.
+
+The BatchProcessor maintains its logic of batching multiple logs together into a single envelope to avoid multiple HTTP requests.
+
+Hybrid SDKs pass every span down to the native SDKs, which will put every log in their BatchProcessor and its cache.
```

Comment:
With that said, will the batch processor only be served as a queue for sending logs or will also invoke integrations for parsing the logs, like beforelog on the native side?

Are there hybrid SDKs using the batch processor at the moment?